runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnRunnerWithKeyedInternals.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.beam.runners.core.StatefulDoFnRunner;
+import org.apache.beam.runners.core.StepContext;
+import org.apache.beam.runners.core.TimerInternals;
+import org.apache.beam.runners.core.TimerInternalsFactory;
+import org.apache.beam.runners.samza.metrics.DoFnRunnerWithMetrics;
+import org.apache.beam.runners.samza.metrics.SamzaMetricsContainer;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.state.TimeDomain;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignature;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignatures;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.joda.time.Instant;
+
+/**
+ * This class wraps a DoFnRunner with keyed StateInternals and TimerInternals access.
+ */
+public class DoFnRunnerWithKeyedInternals<InputT, OutputT> implements DoFnRunner<InputT, OutputT> {
+  private final DoFnRunner<InputT, OutputT> underlying;
+  private final KeyedInternals keyedInternals;
+
+  public static <InputT, OutputT> DoFnRunner<InputT, OutputT> of(
+      PipelineOptions options,
+      DoFn<InputT, OutputT> doFn,
+      SideInputReader sideInputReader,
+      DoFnRunners.OutputManager outputManager,
+      TupleTag<OutputT> mainOutputTag,
+      List<TupleTag<?>> additionalOutputTags,
+      StateInternalsFactory<?> stateInternalsFactory,
+      TimerInternalsFactory<?> timerInternalsFactory,
+      WindowingStrategy<?, ?> windowingStrategy,
+      SamzaMetricsContainer metricsContainer,
+      String stepName) {
+
+    final DoFnSignature signature = DoFnSignatures.getSignature(doFn.getClass());
+    final KeyedInternals keyedInternals;
+    final TimerInternals timerInternals;
+    final StateInternals stateInternals;
+
+    if (signature.usesState()) {
+      keyedInternals = new KeyedInternals(stateInternalsFactory, timerInternalsFactory);
+      stateInternals = keyedInternals.stateInternals();
+      timerInternals = keyedInternals.timerInternals();
+    } else {
+      keyedInternals = null;
+      stateInternals = stateInternalsFactory.stateInternalsForKey(null);
+      timerInternals = timerInternalsFactory.timerInternalsForKey(null);
+    }
+
+    final DoFnRunner<InputT, OutputT> doFnRunner = DoFnRunners.simpleRunner(
+        options,
+        doFn,
+        sideInputReader,
+        outputManager,
+        mainOutputTag,
+        additionalOutputTags,
+        createStepContext(stateInternals, timerInternals),
+        windowingStrategy);
+
+    final DoFnRunner<InputT, OutputT> doFnRunnerWithMetrics = DoFnRunnerWithMetrics
+        .wrap(doFnRunner, metricsContainer, stepName);
+
+    if (keyedInternals != null) {
```

Comment:
This conditional goes right along with the one above to make just two code paths: the one where this method actually does something, and the one that is basically pass through.